utility_network/perform-valve-isolation-trace/src/main/java/com/esri/samples/perform_valve_isolation_trace/PerformValveIsolationTraceController.java

```diff
@@ -240,64 +237,59 @@ private void handleTraceClick() {
 
       // run the trace and get the result
-      ListenableFuture<List<UtilityTraceResult>> utilityTraceResultsFuture = utilityNetwork.traceAsync(utilityTraceParameters);
-      utilityTraceResultsFuture.addDoneListener(() -> {
-        try {
-          List<UtilityTraceResult> utilityTraceResults = utilityTraceResultsFuture.get();
-
-          if (utilityTraceResults.get(0) instanceof UtilityElementTraceResult) {
-            var utilityElementTraceResult = (UtilityElementTraceResult) utilityTraceResults.get(0);
+      utilityNetwork.traceAsync(utilityTraceParameters).toCompletableFuture().whenComplete(
+        (utilityTraceResults, ex) -> {
+          if (ex == null) {
+            if (utilityTraceResults.get(0) instanceof UtilityElementTraceResult) {
+              var utilityElementTraceResult = (UtilityElementTraceResult) utilityTraceResults.get(0);
 
-            if (!utilityElementTraceResult.getElements().isEmpty()) {
+              if (!utilityElementTraceResult.getElements().isEmpty()) {
 
-              // iterate through the map's feature layers
-              mapView.getMap().getOperationalLayers().forEach(layer -> {
-                if (layer instanceof FeatureLayer) {
+                // iterate through the map's feature layers
+                mapView.getMap().getOperationalLayers().forEach(layer -> {
+                  if (layer instanceof FeatureLayer) {
 
-                  // create query parameters to find features whose network source name matches the layer's feature
-                  // table name
-                  var queryParameters = new QueryParameters();
-                  utilityElementTraceResult.getElements().forEach(utilityElement -> {
+                    // create query parameters to find features whose network source name matches the layer's feature
+                    // table name
+                    var queryParameters = new QueryParameters();
+                    utilityElementTraceResult.getElements().forEach(utilityElement -> {
 
-                    String networkSourceName = utilityElement.getNetworkSource().getName();
-                    String featureTableName = ((FeatureLayer) layer).getFeatureTable().getTableName();
-
-                    if (networkSourceName.equals(featureTableName)) {
-                      queryParameters.getObjectIds().add(utilityElement.getObjectId());
-                    }
-                  });
+                      String networkSourceName = utilityElement.getNetworkSource().getName();
+                      String featureTableName = ((FeatureLayer) layer).getFeatureTable().getTableName();
 
-                  // select features that match the query
-                  ListenableFuture<FeatureQueryResult> featureQueryResultListenableFuture =
-                          ((FeatureLayer) layer).selectFeaturesAsync(queryParameters, FeatureLayer.SelectionMode.NEW);
+                      if (networkSourceName.equals(featureTableName)) {
```

Comment:
You can maybe filter and map here.

Reply:
Yeah I think a filter and a map works quite nicely here: 
```java
                        // populate the query parameters with the object IDs of any utility element in the trace result whose network source is equal to the layer's feature table
                       queryParameters.getObjectIds().addAll(utilityElementTraceResult.getElements().stream()
                          .filter(utilityElement ->
                              utilityElement.getNetworkSource().getName().equals(layer.getFeatureTable().getTableName()))
                          .map(UtilityElement::getObjectId).toList());
```